yesod-core/src/Yesod/Core/Class/Yesod.hs

```diff
@@ -70,6 +73,17 @@ class RenderRoute site => Yesod site where
     approot :: Approot site
     approot = guessApproot
 
+    -- | @since 1.6.23.2
+    --   Should we catch an exception, or rethrow it.
+    --   Rethrowing an exception lets the webserver deal with it
+    --   (usually warp).
+    --   catching allows yesod to render the error page.
+    --   the default 'defaultCatchBehavior' is to catch everything
+    --   (even async), except for the
+    --   'Warp.ConnectionClosedByPeer' constructor.
+    catchBehavior :: Proxy site -> SomeException -> CatchBehavior
```

Comment:
This may be too limiting an API. We've often run into situations later where people say they want to be able to change behavior at runtime. What do you think of something like:

```haskell
catchBehavior :: site -> SomeException -> IO CatchBehavior
```

Reply:
Also, we may want to change the default behavior back to what it used to be (not catching async exceptions).

Reply:
@pbrisbin  thoughts?

Reply:
Hmm. Well this is tricky because I'm coming at it from a couple of different perspectives.

As a user of Yesod _on behalf of my users_, I'd agree that I'd **prefer** if the normal error handling and error page triggered on `ThreadKilled` and `AllocationLimitExceeded`. Because, why not?

As an operator of Yesod, I'm not sure if I'd hold that preference for "most async exceptions", because I think there are a lot of async exceptions that are normal in the context of handling web requests, don't ever reach or impact end users, and I don't want them inflating error metrics as if they had.

As someone familiar with the implementation inolved, I don't agree that I'd **expect** the normal error handling and error page triggered on these two exceptions in particular. In my head, the scope of handling a web request is

```
Warp
  Wai
   |----------|
   | Yesod    |
   |   My App |
   |----------|
```

In the case of these two exceptions in particular, I'd expect everything I've boxed there to be dead, like `kill -9` dead. And so I personally wouldn't expect any error recovery to be possible at that level for such exceptions.

Knowing what I've learned recently about async exceptions, I think I could safely apply this expectation to "most async exceptions". I think they all represent `kill -9`-like messaging and so there should be no recovery.

Reply:
so I guess for most users this default of catching everything except timeout & ConnectionClosedByPeer constructor is sensible, but as an operator you could still override this behavior? In other words, everyone is happy?

Reply:
As long as we end up in a place where `TimeoutError` behaves correctly, I'm happy. If we also end up in a place where it's easy for me to say "don't catch any async exceptions", I'm very happy.

I think "don't catch async exceptions" is a more sensible default, for the reasons I tried to describe above, but it's subjective and I don't mind if you go the other way as you're writing the code and Michael can choose to merge it. I just don't see either of us convincing the other, but I don't think we need to.

Reply:
I added a test to verify by default we don't catch `Timeout` either: https://github.com/yesodweb/yesod/pull/1772/files#diff-bd8750d9a236376b742e82438d8e92aeeb76f1d55827fc85c21154c5c4e8f9a0R378

I think that's what you meant, right?

There is a rethrow async function as well (although not as well tested): https://github.com/yesodweb/yesod/pull/1772/files#diff-1fdc39d8a397f906d8f4f69a087048f3432df0b1fd2238ee5459ff1e89bb0edfR655

Reply:
> I think "don't catch async exceptions" is a more sensible default

I think I lean towards this way of thinking, I was already nervous about the original approach since it violates what I generally recommend for async exception handling. I'd be tempted to call the previous change in behavior a bug (since it silently broke other people's code) and go back to the original default, still exposing this configuration point.

To fully flesh out the thinking here, I see four points in the design space for defaults:

1. Rethrow all async exceptions (original behavior)
2. Rethrow a specific list of async exceptions (current PR)
3. Catch a specific list of async exceptions (not proposed yet)
4. Catch all async exceptions (current behavior)

I'm not convinced (2) is a good compromise point, since including Timeout may seem like a good idea here, but honestly seems arbitrary to me. It's entirely possible there's someone else out there using a different async exception for middlewares that we're not aware of.

Reply:
both choices are arbitrary, in anycase I'll rewrite it to the 'rethrow async exception' behavior as default. having this configurable is good enough for me.